### PR TITLE
refactor(expr): generate build-from-prost with procedural macros

### PR DESCRIPTION
**Pull request by wangrunji0408**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

Inspired by user-defined functions, this PR proposes a new expression framework based on procedural macros, so that we can develop built-in functions as easy as UDFs.

In the new framework, a scalar function can be defined in just a few lines:

```rust
#[function("add(int32, int32) -> int32")]
fn add_i32(x: i32, y: i32) -> i32 {
    x + y
}
```

The procedural macro will generate the build-from-prost function according to the signature. It will also generate a function descriptor and add it to a global registry on initialization. This way, we can avoid inconsistencies between function metadata and implementation.

This macro also has some advanced features to handle various situations. Take the generic add function as an example:

```rust
#[function("add(*number, *number) -> auto")]
#[function("add(interval, interval) -> interval")]
pub fn general_add<T1, T2, T3>(l: T1, r: T2) -> Result<T3>
where
    T1: Into<T3> + Debug,
    T2: Into<T3> + Debug,
    T3: CheckedAdd<Output = T3>,
{
    general_atm(l, r, |a, b| {
        a.checked_add(b).ok_or(ExprError::NumericOutOfRange)
    })
}
```

The macro is capable to:
1. allow multiple definations on a generic function.
2. allow wildcards and auto inference in types to expand into multiple definations.
  e.g. `add(*number, *number) -> auto` will be expanded to:
    - `add(int16, int16) -> int16`
    - `add(int16, int32) -> int32`
    - `add(int32, int16) -> int32`
    - ...
4. allow arguments in Rust functions to be either `T`, `Option<T>` or `&mut dyn Write`, and return type to be either `T`, `Option<T>`, `Result<T>` or `Result<Option<T>>`.
  where:
    - `fn(T) -> T` implies returns-null-on-null-input and infallible.
    - `fn(T) -> Result<T>` implies fallible.
    - `fn(Option<T>) -> T` implies called-on-null-input.
    - `fn(Option<T>) -> Result<Option<T>>` is the most generic form.
    - `fn(&mut dyn Write)` implies returning varchar.
6. generate optimal implementation based on the signature.
  e.g. `fn(i32, i32) -> i32` can be applied to array with SIMD acceleration, while `fn(i32, i32) -> Result<i32>` can not.

We expect a lot of boilerplate code to be removed upon completion.
Any comments are welcome!

## Checklist For Contributors

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] ~~I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).~~
- [x] I have demonstrated that backward compatibility is not broken by breaking changes and created issues to track deprecated features to be removed in the future. (Please refer to the issue)
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Checklist For Reviewers

- [ ] I have requested macro/micro-benchmarks as this PR can affect performance substantially, and the results are shown.

## Documentation

- [x] My PR **DOES NOT** contain user-facing changes.

<details><summary>Click here for Documentation</summary>

### Types of user-facing changes

Please keep the types that apply to your changes, and remove the others.

- Installation and deployment
- Connector (sources & sinks)
- SQL commands, functions, and operators
- RisingWave cluster configuration changes
- Other (please specify in the release note below)

### Release note

</details>

### CODE REVIEW

File: src/expr/src/vector_op/to_char.rs [hunk 1]
```diff
@@ -20,7 +20,7 @@ use chrono::format::StrftimeItems;
 use ouroboros::self_referencing;
 use risingwave_common::types::NaiveDateTimeWrapper;
 
-use crate::Result;
+// use risingwave_expr_macro::function;
```
Comment (fuyufjh) [inline src/expr/src/vector_op/to_char.rs:23]:
```suggestion
```

File: src/expr/src/expr/expr_array_to_string.rs
Comment (fuyufjh) [inline src/expr/src/expr/expr_array_to_string.rs:1]: Looks much better than before. Super cool!

File: src/expr/src/vector_op/array_access.rs [hunk 2]
```diff
@@ -14,22 +14,22 @@
...
-        (Some(list), Some(index)) if index > 0 => {
-            let datumref = list.value_at(index as usize)?;
-            if let Some(scalar) = datumref.to_owned_datum() {
-                Ok(Some(scalar.try_into()?))
-            } else {
-                Ok(None)
-            }
-        }
-        _ => Ok(None),
+// FIXME: don't know why this is not working
```
[the remark this answers is not in the document]
Comment (wangrunji0408) [reply src/expr/src/vector_op/array_access.rs]: I observed a weird problem. Functions defined here will not take effect once there are too many definitions.

Specifically:
```rust
// no attribute
pub fn array_access<T: Scalar>(list: ListRef<'_>, index: i32) -> Result<Option<T>> {...}
// 815 function signatures loaded.
```

```rust
#[function("array_access(list, int32) -> boolean")]
#[function("array_access(list, int32) -> int16")]
#[function("array_access(list, int32) -> int32")]
#[function("array_access(list, int32) -> int64")]
#[function("array_access(list, int32) -> float32")]
pub fn array_access<T: Scalar>(list: ListRef<'_>, index: i32) -> Result<Option<T>> {...}
// 820 function signatures loaded.
```

```rust
#[function("array_access(list, int32) -> boolean")]
#[function("array_access(list, int32) -> int16")]
#[function("array_access(list, int32) -> int32")]
#[function("array_access(list, int32) -> int64")]
#[function("array_access(list, int32) -> float32")]
#[function("array_access(list, int32) -> float64")]
pub fn array_access<T: Scalar>(list: ListRef<'_>, index: i32) -> Result<Option<T>> {...}
// 815 function signatures loaded.
// all registration failed
```

More strangely, this only occurs in simulation on macOS. Linux and playground on macOS work well. Given that function registration depends on [ctor](https://github.com/mmastrac/rust-ctor), I guess the problem is caused by link-time optimization. Still debugging...

File: src/stream/src/executor/filter.rs [hunk 3]
```diff
@@ -234,15 +234,16 @@ mod tests {
         };
         let source = MockSource::with_chunks(schema, PkIndices::new(), vec![chunk1, chunk2]);
 
-        let left_expr = InputRefExpression::new(DataType::Int64, 0);
-        let right_expr = InputRefExpression::new(DataType::Int64, 1);
-        let test_expr = new_binary_expr(
-            Type::GreaterThan,
+        let test_expr = build(
```
Comment (BugenZhao) [inline src/stream/src/executor/filter.rs:237]: I guess this PR can also enable us for https://github.com/risingwavelabs/risingwave/issues/7881? 🚀

**Review by BugenZhao — APPROVED** [review]
This PR looks really exciting! 🚀🚀🚀

File: src/stream/src/executor/filter.rs (hunk 3 above)
Comment (wangrunji0408) [reply src/stream/src/executor/filter.rs:237]: Not exactly this PR did, but the ideas behind them are similar. Based on my experience with egg, I feel that a pretty construction of expression is super useful and not hard to implement. I'll have a try on it.

File: src/expr/src/vector_op/array_access.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (wangrunji0408) [reply src/expr/src/vector_op/array_access.rs]: Problem solved after setting `incremental = false` 🥵

[the remark this answers is not in the document]
Comment (wangrunji0408) [reply src/expr/src/vector_op/array_access.rs]: Fixed in 4d306f3. It seems that the default linker on macOS (ld?) caused this problem. When switched to `ld64.lld` (this was configured in `.cargo/config` but is overridden by `RUSTFLAGS` in Makefile.toml) the problem disappears.

**Comment by wangrunji0408** [comment]
@xiangjinwu PTAL 65a83b6

This PR adds `list < list` to the `FUNC_SIG_MAP`, causing a panic on:
https://github.com/risingwavelabs/risingwave/blob/24ecd4d633cc7d11cd80f4cba73c58b7918ba4a2/src/frontend/src/expr/type_inference/func.rs#L92
```
thread '<unnamed>' panicked at 'Functions returning struct or list can not be inferred. Please use `FunctionCall::new_unchecked`.', src/common/src/types/mod.rs:199:47
```

65a83b6 is a workaround for this.

**Pull request merged**